cpp_linter/rest_api/github_api.py

```diff
@@ -67,6 +78,58 @@ def set_exit_code(
             checks_failed, format_checks_failed, tidy_checks_failed
         )
 
+    @staticmethod
+    def _rate_limit_exceeded(
+        response_headers: Union[CaseInsensitiveDict, Dict[str, str]],
+    ):
+        logger.error("RATE LIMIT EXCEEDED!")
+        if "x-ratelimit-reset" in response_headers:
+            reset_epoch = time.gmtime(int(response_headers["x-ratelimit-reset"]))
+            logger.error(
+                "Github REST API rate limit resets on %s",
+                time.strftime("%d %B %Y %H:%M +0000", reset_epoch),
+            )
+        sys.exit(1)
+
+    def api_request(
+        self,
+        url: str,
+        method: Optional[str] = None,
+        data: Optional[str] = None,
+        headers: Optional[Dict[str, Any]] = None,
+    ) -> requests.Response:
+        if self._rate_limit_back_step >= 5 or self._rate_limit_remaining == 0:
+            self._rate_limit_exceeded({})
+        response = self.session.request(
+            method=method or ("GET" if data is None else "POST"),
+            url=url,
+            headers=headers,
+            data=data,
+        )
+        self._rate_limit_remaining = int(
+            response.headers.get("x-ratelimit-remaining", "-1")
+        )
+        log_response_msg(response)
+        if response.status_code in [403, 429]:  # rate limit exceeded
+            # secondary rate limit handling
+            if "retry-after" in response.headers:
+                wait_time = (
+                    float(cast(str, response.headers.get("retry-after")))
+                    * self._rate_limit_back_step
+                )
+                logger.warning(
+                    "SECONDARY RATE LIMIT HIT! Backing off for %f seconds",
+                    wait_time,
+                )
+                time.sleep(wait_time)
+                self._rate_limit_back_step += 1
+                return self.api_request(url, method=method, data=data, headers=headers)
+            # primary rate limit handling
+            if self._rate_limit_remaining == 0:
+                self._rate_limit_exceeded(response.headers)
```

Comment:
This code is executed immediately after a request response has been made. the repsonse's body could still be usable even if there are no allowed requests remaining.

Remove this check in favor of the same check located at the beginning of this function. In doing so, we can combine the `_rate_limit_exceeded()` code into this function because it would only get called from one location.